### PR TITLE
Add search bar and sort by

### PR DESCRIPTION
**Pull request by ThomasOliverKimble**
**Description**
Added new features:
- [x]  Search bar (updates lists in real time, message if no results)
- [x]  'Sort by' functionality allowing to sort the projects by one of the table columns
- [x]  Added a view count to simulations

### CODE REVIEW

File: httpdocs/js/project.js [hunk 1]
```diff
@@ -112,6 +112,9 @@ export default class Project extends User {
       (data ? data.version : this.findGetParameter('version'));
     const src = 'https://cyberbotics.com/wwi/' + version + '/WebotsView.js';
 
+    if (!data)
+      that._updateSimulationViewCount(url);
```
Comment (BenjaminDeleze) [inline httpdocs/js/project.js:116]: Is it useful?
It seems like the returned data is not handled

File: httpdocs/js/webots-cloud.js [hunk 2]
```diff
@@ -719,6 +1005,16 @@ document.addEventListener('DOMContentLoaded', function() {
...
+            document.getElementById('simulation-search-input').value = searchString;
+            /* if (sortBy && sortBy !== 'default') {
+              const columnTitle = document.getElementById('simulation-sort-' + sortBy.split('-')[0]);
+              columnTitle.querySelector('.sort-icon').style.display = 'inline';
+              if (sortBy.split('-')[1] === 'asc') {
+                const sortIcon = columnTitle.querySelector('.sort-icon');
+                sortIcon.classList.toggle('fa-sort-down');
+                sortIcon.classList.toggle('fa-sort-up');
+              }
+            } */
```
Comment (BenjaminDeleze) [inline httpdocs/js/webots-cloud.js]: Is there a reason to keep the commented part or can we remove it?

File: httpdocs/js/webots-cloud.js [hunk 3]
```diff
@@ -401,6 +565,71 @@ document.addEventListener('DOMContentLoaded', function() {
...
+              sortIcon.classList.toggle('fa-sort-up');
+            }
+            sort += '-desc';
+          } else
+            sort += '-desc';
+
+          title.querySelector('.sort-icon').style.display = 'inline';
+          setSorts(type, sort);
+          searchAndSortTable(type);
+        })
```
Comment (BenjaminDeleze) [inline httpdocs/js/webots-cloud.js]:
```suggestion
        });
```

File: httpdocs/js/webots-cloud.js [hunk 4]
```diff
@@ -401,6 +565,71 @@ document.addEventListener('DOMContentLoaded', function() {
...
+        document.getElementById(type + '-search-input').addEventListener('keyup', function(event) {
+          if (!getSearch('delay')) {
+            setSearches('delay', true);
+            setTimeout(() => {
+              setSearches(type, document.getElementById(type + '-search-input').value);
+              setPages(type, 1);
+              updateSearchIcon(type);
+              searchAndSortTable(type);
+              setSearches('delay', false);
+            }, '300')
```
Comment (BenjaminDeleze) [inline httpdocs/js/webots-cloud.js]:
```suggestion
            }, '300');
```

File: httpdocs/js/webots-cloud.js [hunk 5]
```diff
@@ -415,7 +644,17 @@ document.addEventListener('DOMContentLoaded', function() {
...
+          sort = getSort(activeTab);
+          search = getSearch(activeTab);
+          let url = new URL(document.location.origin + '/' + activeTab);
+          if (getPage(activeTab) !== 1)
+            url.searchParams.append('p', getPage(activeTab));
+          if (getSort(activeTab) && getSort(activeTab) !== 'default')
+            url.searchParams.append('sort', getSort(activeTab));
+          if (getSearch(activeTab) && getSearch(activeTab) !== '')
+            url.searchParams.append('search', getSearch(activeTab));
+          updateSearchIcon(activeTab)
```
Comment (BenjaminDeleze) [inline httpdocs/js/webots-cloud.js]:
```suggestion
          updateSearchIcon(activeTab);
```

File: httpdocs/js/webots-cloud.js [hunk 6]
```diff
@@ -428,6 +667,44 @@ document.addEventListener('DOMContentLoaded', function() {
...
+          searchIcon.classList.add('fa-xmark');
+        } else if (searchIcon.classList.contains('fa-xmark') && getSearch(type).length === 0) {
+          searchIcon.classList.add('fa-search');
+          searchIcon.classList.remove('fa-xmark');
+        }
+      } else {
+        updateSearchIcon('scene');
+        updateSearchIcon('animation');
+        updateSearchIcon('simulation');
+        return;
```
Comment (BenjaminDeleze) [inline httpdocs/js/webots-cloud.js]:
```suggestion
```
Unnecessary return as we reach the end of the function

File: httpdocs/js/webots-cloud.js [hunk 7]
```diff
@@ -661,18 +936,25 @@ document.addEventListener('DOMContentLoaded', function() {
       });
     }
 
-    function listAnimations(type, page) {
+    function listAnimations(type, page, sortBy, searchString) {
       const typeName = (type === 'A') ? 'animation' : 'scene';
       const capitalizedTypeName = typeName.charAt(0).toUpperCase() + typeName.slice(1);
       const offset = (page - 1) * pageLimit;
-      fetch('/ajax/animation/list.php', {method: 'post', body: JSON.stringify({offset: offset, limit: pageLimit, type: type})})
+      fetch('/ajax/animation/list.php', {method: 'post',
```
Comment (BenjaminDeleze) [inline httpdocs/js/webots-cloud.js]:
```suggestion
      fetch('/ajax/animation/list.php', {method: 'post',
```

File: httpdocs/js/webots-cloud.js [hunk 8]
```diff
@@ -206,7 +271,7 @@ document.addEventListener('DOMContentLoaded', function() {
       const admin = project.email ? project.email.endsWith('@cyberbotics.com') : false;
       const words = data.url.substring(19).split('/');
       const dotIndex = data.url.lastIndexOf('/') + 1;
-      const thumbnailUrl = (data.url.slice(0, dotIndex) + "." + data.url.slice(dotIndex)).replace('github.com', 'raw.githubusercontent.com').replace('/blob', '').replace('.wbt', '.jpg');
+      const thumbnailUrl = (data.url.slice(0, dotIndex) + '.' + data.url.slice(dotIndex)).replace('github.com', 'raw.githubusercontent.com').replace('/blob', '').replace('.wbt', '.jpg');
```
Comment (BenjaminDeleze) [inline httpdocs/js/webots-cloud.js]: line too long

**Review by BenjaminDeleze — CHANGES_REQUESTED** [review]
Works really well!

File: httpdocs/js/project.js (hunk 1 above)
Comment (ThomasOliverKimble) [reply httpdocs/js/project.js:116]: If `data` is set, it is either an animation or a scene, so here all we do is call an iterator to update the simulation view count for that particular url (simulation) in the database.

File: httpdocs/js/webots-cloud.js (hunk 2 above)
Comment (ThomasOliverKimble) [reply httpdocs/js/webots-cloud.js]: My bad! it has been removed.

**Review by BenjaminDeleze — APPROVED** [review]
Thank you

**Pull request merged**